File: dalcom/boj/11505/11505.cpp
#include <bits/stdc++.h>
using namespace std;
#define DIV 1000000007

vector<int> arr(1000001);
vector<long> tree(1000001, 1);

long init(int start, int end, int index) {
    if (start == end) {
        return tree[index] = arr[start];
    }

    int mid = (start + end) / 2;
    return tree[index] = (init(start, mid, index * 2) % DIV) *
                         (init(mid + 1, end, index * 2 + 1) % DIV);
}

long mul(int start, int end, int index, int left, int right) {
    if (left > end || right < start) {
        return 1;
    }

    if (left <= start && end <= right) {
        return tree[index];
    }

    int mid = (start + end) / 2;
    return (mul(start, mid, index * 2, left, right) % DIV) *
           (mul(mid + 1, end, index * 2 + 1, left, right) % DIV);
}

long long update(int start, int end, int treeIdx, int arrIdx, long long val) {
    if (arrIdx < start || arrIdx > end) {
        return tree[treeIdx];
    }

    if (start == end) {
        return tree[treeIdx] = val;
    }

    int mid = (start + end) / 2;
    tree[treeIdx] = (update(start, mid, treeIdx * 2, arrIdx, val) % DIV) *
                    (update(mid + 1, end, treeIdx * 2 + 1, arrIdx, val) % DIV);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m, k;

    cin >> n >> m >> k;

    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    init(0, n - 1, 1);

    for (int i = 0; i < m + k; i++) {
        int a, b, c;
        cin >> a >> b >> c;
        b = b - 1;

        if (a == 1) {
            arr[b] = c;
            update(0, n - 1, 1, b, c);
        } else if (a == 2) {
            c = c - 1;
            cout << mul(0, n - 1, 1, b, c) % DIV << '\n';
        }
    }

    long long a = 6, b = 4;
    cout << b / a;

    return 0;
}

File: dalcom/boj/11505/11505.py
from sys import stdin as input

input = open("/Volumes/Program/ps-study/dalcom/11505/input.txt", "rt")

n, m, k = map(int, input.readline().split())

arr = []

for _ in range(n):
    arr.append(int(input.readline()))

tree = [0 for _ in range(len(arr) * 4 + 1)]


def init(start, end, index):
    if start == end:
        tree[index] = arr[start]
        return tree[index]

    mid = (start + end) // 2
    tree[index] = (
        init(start, mid, index * 2) * init(mid + 1, end, index * 2 + 1) % 1000000007
    )
    return tree[index]


def mul(start, end, index, left, right):
    if left > end or right < start:
        return 1

    if left <= start and right >= end:
        return tree[index]

    mid = (start + end) // 2
    return (
        mul(start, mid, index * 2, left, right)
        * mul(mid + 1, end, index * 2 + 1, left, right)
        % 1000000007
    )


def update(start, end, index, arr_index, diff):
    if arr_index < start or arr_index > end:
        return

    if start == end:
        tree[index] = diff
        return

    mid = (start + end) // 2
    update(start, mid, index * 2, arr_index, diff)
    update(mid + 1, end, index * 2 + 1, arr_index, diff)
    tree[index] = tree[index * 2] * tree[index * 2 + 1] % 1000000007


init(0, len(arr) - 1, 1)

for _ in range(m + k):
    a, b, c = map(int, input.readline().split())
    b = b - 1
    if a == 1:
        arr[b] = c
        update(0, len(arr) - 1, 1, b, c)
    elif a == 2:
        c = c - 1
        print(int(mul(0, len(arr) - 1, 1, b, c)) % 1000000007)
